cmd/geth:Implement freezer truncation as a subcommand(fixes #31135)

sivaratrisrinivas opened the pull request:
Hii, I have implemented a new subcommand `geth db truncate-freezer` that truncates the freezer at the merge block, keeping headers but removing bodies. This addresses issue #31135.

## These are the Implementation Details

- Added a new subcommand `truncate-freezer` to the `db` command
- Consolidated flags into reusable variables
- Implemented the truncation logic that:
  - Finds the merge block using binary search
  - Preserves headers and hashes before truncation
  - Truncates all tables
  - Re-inserts the preserved headers and hashes

This implementation follows the same pattern as the existing `prune-history` command and uses the same underlying `truncateAncientStore` method.

Fixes #31135

MariusVanDerWijden commented on cmd/geth/dbcmd.go:
```diff
@@ -929,3 +982,345 @@ func inspectHistory(ctx *cli.Context) error {
+	)
+
+	log.Info("Searching for merge block using binary search", "head", *headNumber)
+
+	for low <= high {
+		mid := (low + high) / 2
+		header := rawdb.ReadHeader(db, rawdb.ReadCanonicalHash(db, mid), mid)
+		if header == nil {
+			return fmt.Errorf("header %d not found", mid)
+		}
+
+		if header.Difficulty.Sign() == 0 {
+			// This is a post-merge block, look earlier
+			high = mid - 1
+			mergeBlock = mid
+			found = true
+		} else {
+			// This is a pre-merge block, look later
+			low = mid + 1
+		}
+	}
```
You could do something like this here
```suggestion
    sort.Search(*headNumber, func(index int) bool {
         header := rawdb.ReadHeader(db, rawdb.ReadCanonicalHash(db, index), index)
         if header == nil {
             panic(fmt.Sprintf("header %d not found", index))
         }
         return header.Difficulty.Sign() == 0
         })
```

jwasinger replied:
Can't we just assume the merge block is hardcoded, and assume that we are running on a pre-merged chain?

s1na commented on cmd/geth/dbcmd.go:
```diff
@@ -929,3 +982,345 @@ func inspectHistory(ctx *cli.Context) error {
+
+		// First, read all headers up to the merge block
+		log.Info("Reading headers to preserve them", "count", mergeBlock)
+		headers := make([][]byte, mergeBlock)
```
So this will be 15 Gbs of memory on mainnet

s1na commented on cmd/geth/dbcmd.go:
```diff
@@ -929,3 +982,345 @@ func inspectHistory(ctx *cli.Context) error {
+		}
+
+		// Re-insert the headers and hashes
+		log.Info("Re-inserting headers", "count", len(headers))
```
I think this will not work. Yes you can re-insert the headers, but the freezer has the assumption that all of its tables have the same length. So upon next boot it will truncate the headers to match that of blocks, receipts etc.

s1na replied:
It seems we have two options:

- relax this assumption in the freezer
- Move the headers and hashes to a fresh freezer

sivaratrisrinivas replied:
Thankyou for pointing out the issues.

sivaratrisrinivas commented:
I've updated the PR with optimizations that address all the concerns raised:

### 1. Memory Usage Issue
- Implemented batch processing with a configurable batch size (default: 10,000)
- Added progress reporting to show status during processing
- Extracted batch processing logic into dedicated helper functions

### 2. Freezer Table Length Assumption
- Created a temporary freezer specifically for headers and hashes
- Completely truncate the original freezer (to zero) before re-inserting headers
- Added proper error handling when copying between freezers

### 3. Hardcoded Merge Block Values
- Added `useHardcodedMergeFlag` flag
- Created `knownMergeBlocks` map with network-specific merge block numbers
- Extracted merge block detection into a dedicated function

### Additional Improvements
- **Better Code Organization**: Broke down the large function into smaller, focused helper functions
- **Enhanced User Experience**: Added progress reporting for long-running operations
- **Improved Error Handling**: Using `%w` for error wrapping
- **Configurability**: Made batch size configurable via command-line flag

Please let me know your thoughts on it.

sivaratrisrinivas commented:
Hii @s1na , I fixed the issues you have mentioned and made some modifications to the codebase to make it more readable and maintainable. Can you please review it and share your feedback.

fjl commented:
We decided to work on our own version of this. Thank you for your contribution!
